sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java

```diff
@@ -145,15 +145,109 @@ private CosmosPagedFluxOptions createCosmosPagedFluxOptions() {
+        final CosmosDiagnosticsContext cosmosCtx = ctxAccessor.create(
+            pagedFluxOptions.getSpanName(),
+            pagedFluxOptions.getAccountTag(),
+            BridgeInternal.getServiceEndpoint(pagedFluxOptions.getCosmosAsyncClient()),
+            pagedFluxOptions.getDatabaseId(),
+            pagedFluxOptions.getContainerId(),
+            pagedFluxOptions.getResourceType(),
+            pagedFluxOptions.getOperationType(),
+            pagedFluxOptions.getOperationId(),
+            pagedFluxOptions.getEffectiveConsistencyLevel(),
+            pagedFluxOptions.getMaxItemCount(),
+            pagedFluxOptions.getDiagnosticsThresholds(),
+            null);
+
+        ctxAccessor.setSamplingRateSnapshot(cosmosCtx, pagedFluxOptions.getSamplingRateSnapshot());
+
+        Flux<FeedResponse<T>> result = tracerProvider
+            .runUnderSpanInContext(publisher)
+            .doOnEach(signal -> {
+                FeedResponse<T> response = signal.get();
+                Context traceCtx = DiagnosticsProvider.getContextFromReactorOrNull(signal.getContextView());
+
+                synchronized (lockHolder) {
+                    switch (signal.getType()) {
+                        case ON_COMPLETE:
+                            this.recordFeedResponse(pagedFluxOptions, traceCtx, tracerProvider, response, feedResponseConsumerLatencyInNanos);
+                            tracerProvider.recordFeedResponseConsumerLatency(
+                                signal,
+                                Duration.ofNanos(feedResponseConsumerLatencyInNanos.get()));
+
+                            tracerProvider.endSpan(traceCtx);
+
+                            break;
+                        case ON_NEXT:
+                            this.recordFeedResponse(pagedFluxOptions, traceCtx, tracerProvider, response, feedResponseConsumerLatencyInNanos);
+                            break;
+                        case ON_ERROR:
+                            tracerProvider.recordFeedResponseConsumerLatency(
+                                signal,
+                                Duration.ofNanos(feedResponseConsumerLatencyInNanos.get()));
+
+                            // all info is extracted from CosmosException when applicable
+                            tracerProvider.endSpan(
+                                traceCtx,
+                                signal.getThrowable()
+                            );
+
+                            break;
+
+                        default:
+                            break;
+                    }
+                }
+            });
+
+        return Flux
+            .deferContextual(reactorCtx -> result
+                .doOnCancel(() -> {
+                    Context traceCtx = DiagnosticsProvider.getContextFromReactorOrNull(reactorCtx);
+                    synchronized (lockHolder) {
+                        tracerProvider.endSpan(traceCtx);
```

Comment:
Can you please do me one favor and triple check that exception (non-errors) leaking from traceProvider.endSPan don't cause the Flux to hang in any of the stages where endSpan is called? Annie might have validated it already - but that would be useful to check at least manually.